src/coreclr/jit/lclmorph.cpp

```diff
@@ -995,150 +1120,130 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
         }
         else
         {
-            // TODO-ADDR: Add TYP_STRUCT support to LCL_FLD.
-            return;
+            indirLayout = indir->AsBlk()->GetLayout();
         }
 
-        GenTreeFlags flags = GTF_EMPTY;
-
-        if ((user != nullptr) && user->OperIs(GT_ASG) && (user->AsOp()->gtGetOp1() == indir))
+        // How does the "indir" match the underlying location?
+        //
+        enum class StructMatch
         {
-            flags |= GTF_VAR_DEF | GTF_DONT_CSE;
+            Exact,
+            Compatible,
+            Partial
+        };
 
-            if (indir->OperIs(GT_LCL_FLD))
-            {
-                // Currently we don't generate TYP_STRUCT LCL_FLDs so we do not need to
-                // bother to find out the size of the LHS for "partial definition" purposes.
-                assert(!varTypeIsStruct(indir->TypeGet()));
+        // We're only processing TYP_STRUCT variables now.
+        assert(varDsc->GetLayout() != nullptr);
 
-                if (genTypeSize(indir->TypeGet()) < m_compiler->lvaLclExactSize(val.LclNum()))
-                {
-                    flags |= GTF_VAR_USEASG;
-                }
+        StructMatch match = StructMatch::Partial;
+        if (val.Offset() == 0)
+        {
+            if (indirLayout->GetClassHandle() == varDsc->GetStructHnd())
+            {
+                match = StructMatch::Exact;
+            }
+            else if (ClassLayout::AreCompatible(indirLayout, varDsc->GetLayout()))
+            {
+                match = StructMatch::Compatible;
             }
         }
 
-        indir->gtFlags = flags;
+        // Current matrix of matches/users/types:
+        //
+        // |------------|------|---------|--------|
+        // | STRUCT     | CALL | ASG     | RETURN |
+        // |------------|------|---------|--------|
+        // | Exact      | None | LCL_VAR | None   |
+        // | Compatible | None | LCL_VAR | None   |
+        // | Partial    | None | OBJ     | None   |
+        // |------------|------|---------|--------|
+        //
+        // |------------|------|---------|--------|----------|
+        // | SIMD       | CALL | ASG     | RETURN | HWI/SIMD |
+        // |------------|------|---------|--------|----------|
+        // | Exact      | None | None    | None   | None     |
+        // | Compatible | None | None    | None   | None     |
+        // | Partial    | None | None    | None   | None     |
+        // |------------|------|---------|--------|----------|
+        //
+        // TODO-ADDR: delete all the "None" entries and always
+        // transform local nodes into LCL_VAR or LCL_FLD.
+
+        assert(user->OperIs(GT_ASG) && indir->TypeIs(TYP_STRUCT));
 
-        INDEBUG(m_stmtModified = true;)
+        *pStructLayout = indirLayout;
+
+        if ((match == StructMatch::Exact) || (match == StructMatch::Compatible))
+        {
+            return IndirTransform::LclVar;
+        }
```

Comment:
Do you expect to start using the `StructMatch` for something more complex in the future?